test-framework/junit5-mockito/src/main/java/io/quarkus/test/junit/mockito/internal/CreateMockitoSpiesCallback.java

```diff
@@ -0,0 +1,70 @@
+package io.quarkus.test.junit.mockito.internal;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.inject.Qualifier;
+
+import org.mockito.Mockito;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.InstanceHandle;
+import io.quarkus.test.junit.callback.QuarkusTestBeforeAllCallback;
+import io.quarkus.test.junit.mockito.InjectSpy;
+
+public class CreateMockitoSpiesCallback implements QuarkusTestBeforeAllCallback {
+
+    @Override
+    public void beforeAll(Object testInstance) {
+        Class<?> current = testInstance.getClass();
+        while (current.getSuperclass() != null) {
+            for (Field field : current.getDeclaredFields()) {
+                InjectSpy injectSpyAnnotation = field.getAnnotation(InjectSpy.class);
+                if (injectSpyAnnotation != null) {
+                    Object beanInstance = getBeanInstance(testInstance, field);
+                    Object spy = createSpyAndSetTestField(testInstance, field, beanInstance);
+                    MockitoMocksTracker.track(testInstance, spy, beanInstance);
```

Comment:
I'm afraid that this could get us into trouble.

What this does is basically set a private field on the CDI proxy. For mock that works great because the mock is just some object. But in this case you would be setting the that private field to the CDI proxy itself.
I really think we need to find a more safe way of providing this facility.

Reply:
I am not sure I am following the concern.
In both the mock and spy use cases we are wrapping an Object into the Mockito Framework. 
In the case of a Mock that is that Object is instantiated with null for all internal variables, and the Mockito Framework is set to doReturn(null) for any method that returns anything, or doNothing() for methods that return void.
In the case of the Spy that Object is a fully instantiated Object with all of the variables set how they are expected. Then the Mockito Framework will by default call doCallRealMethod() on all method calls into that Object.

In both cases, we are putting into the CDI proxy a Mockito wrapped Object.

Reply:
The problem is that with the Spy, what you are putting in, actually contains a reference to the object itself, no?

Reply:
Yes, you are just using a wrapper pattern to provide additional functionality to the Object.

At the end, the CDI proxy isn't pointing to the first Object it originally created, it is pointing to the Mockito Framework wrapped Object. If the CDI interacts with that Bean, it will be doing so to the wrapped Object, not to the original Bean itself.

Reply:
So I think I haven't been clear.

When  `MockSupport.installMock()` gets called (down the line) what happens is that whatever you pass as the objects gets stuffed into a special field called `mock` of  the CDI proxy.
In this case, you are passing an object which essentially wraps the proxy itself (unless I have misunderstood). 
So this `mock` field of the CDI proxy would (once the intermediate is removed) essentially contains a reference to itself. That is what has me worried.

Reply:
Okay, I ran some additional tests and compared the memory addresses for the different objects.
In the test I had one Object annotated with @InjectMock and another Object annotated with @InjectSpy. I then ran two @Test method and compared the actual Bean, ClientProxy, and mock/spy objects.

In both @Test cases, the actual CDI Bean instance stayed the same (mock = @34d3bfca and spy = @1d57b8f1), and the ClentProxy instance stayed the same (mock = @16660 and spy = @16705). However, the Mockito wrapped instance change between each test (mock = test1 = @16662, test2 = @17364, and the spy = test1 = @16707, test2 = @17370).

Does this help address your concern?

Reply:
Thanks for the info, I'll take a look at it and get back to you

Reply:
There is a copy of the CDI proxy deep inside the `mock` field. See

![Screenshot from 2020-05-18 18-02-23](https://user-images.githubusercontent.com/4374975/82228508-e0ab6e80-9931-11ea-8f7a-85f2a9890d34.png)

It might be completely safe, but I would like to hear from @mkouba or @manovotn first :)

Reply:
@geoand aren't we doing the same thing in `CreateMockitoMocksCallback#beforeAll()`? When you retrieve the CDI bean there, it also looks to be a proxy, right?

Reply:
I mean you also base the mock on the proxy and then store that into ` MockitoMocksTracker`. I don't know the internals but I would assume there isn't much difference in spy versus mock in this regard.

Reply:
I think that it shoudln't be a problem although it's a bit "tangled". 

There is a difference though because for `@InjectMock` a mock object is created and injected into the test class instance AND this mock is set to the `mock` field of the relevant client proxy object so that any client proxy method invocation is redirected to the mock. 

For `@InjectSpy` a copy (according to the javadoc) of the client proxy object is created AND then this copy is set to the `mock` field of the relevant client proxy object. And it seems to work because the spy copy is created before the `mock` field is set and so if the client proxy is invoked (i.e. from another bean) the spy copy that has `mock=null` is used and so non-stubbed methods delegate to the underlying bean instance.

Reply:
Oh my bad, `Mockito.mock()` is based off a `Class`, not actual instance...
In that case, it might be "cleaner" to unwrap the proxies before creating the spy, WDYT?

Reply:
> In that case, it might be "cleaner" to unwrap the proxies before creating the spy, WDYT?

Just to be clear, you are suggesting we do something like:

`ClientProxyUnwrapper unwrapProxy = new ClientProxyUnwrapper();`
`Object spy = Mockito.spy(unwrapProxy.apply(beanInstance));`

I tested this locally, and the "WithSpiesTest" tests all work with that. This appears to be a good solution, as we are not adding a circular reference to the Proxy.

Reply:
I like @manovotn's idea, WDYT @mkouba ?

Reply:
I'm not sure it's a good idea because if `Mockito.spy()` copies the state and does not delegate to the original instance we may end up with unexpected behavior. Let's say we have `@ApplicationScoped Foo` then `@InjectSpy Foo` -> this creates a copy of Foo instance. Now if you call an unstubbed method on the spy instance the state of the original bean instance does not change. Ideally, test methods should not be connected in any way but since the `@ApplicationScoped` beans are shared...

Reply:
So @mkouba do you propose we just leave the PR as is?

Reply:
Frankly I never used the spy functionality before but looking at the docs, this seems to be what you would expect - that the real instance is in fact *not* invoked, it's state unchanged.

Reply:
FWIW, @mkouba fixed the issue he mentioned above

Reply:
> Ideally, test methods should not be connected in any way but since the `@ApplicationScoped` beans are shared...

I still think we should unwrap the ClientProxy. I did a quick test with the "quarkus-integration-test-injectmock" project and if we just create the Spy on the ClientProxy then the internal state of the actual bean will persist between test. I agree with @mkouba above, that test state should not be shared. If we do share the state, then Testers will see unexpected results.

I updated MessageServiceImpl to:
`private int count = 0;`

    @Override
    public String getMessage() {
        System.out.println("Test Instance: "+this);
        System.out.println("Count: "+count++);
        return "hello";
    }
Between all of the test in the integration project, the MessageServiceImpl is called three times.

Spy on ClientProxy:
`Test Instance: io.quarkus.it.mockbean.MessageServiceImpl@62ce03cc
Count: 0`
`Test Instance: io.quarkus.it.mockbean.MessageServiceImpl@62ce03cc
Count: 1`
`Test Instance: io.quarkus.it.mockbean.MessageServiceImpl@62ce03cc
Count: 2`

Spy on Unwrapped ClientProxy:
`Test Instance: org.mockito.codegen.MessageServiceImpl$MockitoMock$718960814@9131950
Count: 0`
`Test Instance: org.mockito.codegen.MessageServiceImpl$MockitoMock$718960814@38d42ab7
Count: 0`
`Test Instance: io.quarkus.it.mockbean.MessageServiceImpl@2553ddc4
Count: 0`

I will defer and leave the Spy at the ClientProxy if y'all really think that the Test should alter the internal state of ApplicationScoped beans. 
Please advise how y'all would like to proceed.

Reply:
I don't have a strong opinion. Both approaches probably have pros and cons. To me it sounds more natural to delegate the calls to the real instance but the contract of mockito spy is probably more important.

Reply:
I think we should stick to what mockito spy contract is since we are trying to mimick that.
But truth be told, I am not sure I see the use case where you would actually need a spy versus standard mock...

Reply:
I provided a few reasons in the documentation.

An example, in my current project I am working on, we have three possible methods that will all return the same object, but will go about looking up the data for the object different ways. We determine which method to call based on the JSON object passed into the project. When testing, we don't really want to alter how the code operates, we just want to be able to verify that the logic called the correct method.
To do that with a Mock it would have to create the mock and then in setup (@BeforeEach) call doCallRealMethod().when(logic).method() for each public/protected/package private method on the class. Just creating a Spy eliminates that setup and simplifies maintenance as new methods are added.

For the majority of time, a Mock is just fine and will accomplish the majority of testing. Having the ability to use a Spy just provides more options to use when they are beneficial.